### ISSUE TITLE
Test failure in CI build 5381

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#1928]
The following test appears to have failed:

[#5381](https://circleci.com/gh/cockroachdb/cockroach/5381):

```
W0804 02:42:34.576428 451 storage/store.go:584  error gossiping first range data: storage/replica.go:971: node is stopping
W0804 02:42:34.576915 451 storage/store.go:603  error gossiping configs: storage/replica.go:971: node is stopping
I0804 02:42:34.588687 451 storage/replica.go:988  gossiping cluster id  from store 3, range 1
I0804 02:42:34.590486 451 storage/replica.go:996  gossiping sentinel from store 3, range 1
I0804 02:42:34.590932 451 storage/replica.go:1000  gossiping first range from store 3, range 1
panic: test timed out after 5m0s

goroutine 25703 [running]:
testing.func·008()
	/usr/src/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/src/go/src/time/sleep.go:129 +0x59

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x14b7ea0, 0x19e5600, 0xba, 0xba, 0x1)
	/usr/src/go/src/testing/testing.go:556 +0xdc3
--
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:220 +0x525
github.com/cockroachdb/cockroach/util/stop.func·001()
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:79 +0x8c
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:80 +0x14c
FAIL	github.com/cockroachdb/cockroach/storage	300.045s
=== RUN TestBatchBasics
I0804 02:38:31.575296 499 storage/engine/rocksdb.go:99  opening in-memory rocksdb instance
I0804 02:38:31.576836 499 storage/engine/rocksdb.go:124  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.01s)
=== RUN TestBatchGet
I0804 02:38:31.582699 499 storage/engine/rocksdb.go:99  opening in-memory rocksdb instance
I0804 02:38:31.583330 499 storage/engine/rocksdb.go:124  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
I0804 02:38:31.587966 499 storage/engine/rocksdb.go:99  opening in-memory rocksdb instance

```
Please assign, take a look and update the issue accordingly.

**Comment by kkaneda** [comment]
It seems this failure can be reproduced locally by setting `clusterIDGossipTTL` to `1 * time.Millisecond`. 

Added some logging to `redirectOnOrAcquireLeaderLease`. It seems Store 2 got stuck while trying to acquire a lease:

```
  ...
I0804 04:15:19.309798 38932 storage/replica.go:990  gossiping cluster id  from store 3, range 1
I0804 04:15:19.309839 38932 storage/replica.go:990  gossiping cluster id  from store 1, range 1
I0804 04:15:19.309884 38932 storage/replica.go:390  request for leader lease: store=store=1:1 ([]=)
W0804 04:15:19.310125 38932 rpc/server.go:351  rpc: server cannot decode request: read tcp 127.0.0.1:55888: use of closed network connection
W0804 04:15:19.310336 38932 rpc/tls.go:41  listening via tcp to 127.0.0.1:0 without TLS
I0804 04:15:19.310602 38932 storage/replica.go:990  gossiping cluster id  from store 2, range 1
I0804 04:15:19.310646 38932 storage/replica.go:390  request leader lease: store=store=2:2 ([]=)
I0804 04:15:19.310940 38932 storage/replica.go:990  gossiping cluster id  from store 3, range 1
I0804 04:15:19.310986 38932 storage/replica.go:390  request leader lease: store=store=3:3 ([]=)
I0804 04:15:19.311558 38932 storage/replica.go:392  request leader lease: store=store=1:1 ([]=): done
   ...
I0804 04:15:19.312470 38932 storage/replica.go:392  request for leader lease: store=store=3:3 ([]=): done
   ...
```

**Comment by tschottdorf** [comment]
I see "node is stopping" messages, so this looks (to me) like a stopper problem. I already have a local patch that I occasionally use to see what's blocking the shutdown, let me just put that into a PR.

**Comment by tschottdorf** [comment]
The offending goroutine in the trace is here:

```go
	// Send lease request directly to raft in order to skip unnecessary
	// checks from normal request machinery, (e.g. the command queue).
	// Note that the command itself isn't traced, but usually the caller
	// waiting for the result has an active Trace.
	errChan, pendingCmd := r.proposeRaftCommand(r.context(), args)
	if err := <-errChan; err != nil {
		return err
	}
```

so it's one of the cases of "propose to Raft and never hear back". It's highly likely that it's this one:

>
I0804 02:38:34.732137 451 raft/raft.go:390  200000002 became follower at term 6
I0804 02:38:34.732472 451 raft/raft.go:207  newRaft 200000002 [peers: [100000001,300000003], term: 6, commit: 28, applied: 28, lastindex: 28, lastterm: 6]
I0804 02:38:34.732842 451 raft/raft.go:620  200000002 no leader at term 6; dropping proposal

dropping the proposal should be OK because MultiRaft should repropose it when a leader gets elected. That doesn't seem to have happened. @bdarnell?

**Comment by bdarnell** [comment]
MultiRaft will repropose when a leader is elected, but can a leader be elected during shutdown? I think the RunTask in `Server.ServeHTTP` will prevent that. I think we've generally had to execute a dummy command at the end of a test to ensure that a lease is definitely held and won't be refreshed during the draining period.

**Comment by bdarnell** [comment]
Actually, this points to a deeper problem: Any store holding a replica of the first range will have a loop that calls getLeaseForGossip. When that replica is removed, this loop exits on its next iteration, but if it's stuck in a call to getLeaseForGossip nothing will break it out of that state. I think we need to add some sort of fail-fast option to proposeRaftCommand so that we can return an error instead of queueing indefinitely.

**Comment by tschottdorf** [comment]
why is that a deeper problem? If we just fixed Raft so that it would process until the stopper is ready to tear down, shouldn't we be ok? I haven't looked at the `ServeHTTP` RunTask, but I think the idea was to not accept client requests while draining, but still allow internal ones (like Raft progress). Maybe we've just messed something up there.

Also, if we remove a replica, won't we tear down the Raft group, and won't this signal all waiting commands?

**Comment by bdarnell** [comment]
In raft tests we have 3 nodes with 3 stoppers, and we stop the nodes one at a time. By the time the third node is shutting down the other two are already drained.

But you're right; removing the raft group should cancel all waiting commands. The deeper issue is that it's sticking around. The replica of range 1 has been removed from this node but methods on this `Replica` are blocking forever.

The problem is that we have two gossip goroutines, one for the first range, and one for any range that includes config data. Both of them may call `getLeaseForGossip`, and they are serialized by `llMu`. Whichever one runs first proposes a raft command, and has it aborted when the range is removed. The second goroutine is then able to acquire `llMu`, and it proposes a second lease command. This ends up recreating the raft group in `Store.ProposeRaftCommand`. 

I think we need to change `Store.ProposeRaftCommand` to return an error if we try to propose a command for a range that we don't have a replica of.

**Comment by kkaneda** [comment]
Might not be related to the discussion, but I think the test needs to be fixed to correctly wait until range removal. Will send a PR.

**Comment by tamird** [comment]
Looks to be fixed in #1964, the test was just wrong.
